Allow permalinks

LewisAJones opened the pull request:
This PR adds increased functionality to `get_attribution`. It allows users to generate permalinks for collections of UUIDs which can be used to share and document contributors, licenses, etc. This added functionality required the generation of `phy_POST` which I’ve attempted to standardise with `phy_GET`. Happy to make any further changes that might be worthwhile!

Closes #81.

willgearty commented on R/get_attribution.R:
```diff
@@ -130,8 +152,16 @@ get_attribution <- function(uuid = NULL, img = NULL, text = FALSE) {
   # Return data ----------------------------------------------------------
+  # Add permalink?
+  if (permalink) {
+    att$permalink <- perm
```
I like the use of a list here, but should the permalink should be at the same level as the individual images? Maybe the images should be a list item (`att$images`) so that they are collectively at the same level, but individually one level down? Therefore, looping through the list won't cause problems for users (because they would instead always loop through `att$images`, which will always be the same length as `uuids`).

willgearty replied:
If you go with my suggested return behavior change, the text paragraph could also be returned as an item at the top level.

willgearty commented on R/get_attribution.R:
```diff
@@ -18,9 +20,13 @@
 #'   image uuid, license, and license abbreviation. If `text` is set to
 #'   `TRUE`, a text paragraph with the contributor name, year of contribution,
-#'    and license type is returned.
+#'    and license type is returned. If `permalink` is set to `TRUE`, a 
```
Technically, the text paragraph is printed, not "returned".

willgearty reviewed (changes requested):
This all looks and works great, thanks @LewisAJones! I have some comments (below).

I also have a suggestion for the return behavior of `get_attribution`. I realize you haven't fundamentally changed it here, but I think would be good to change as part of this PR. Right now, if `text = TRUE`, the output is printed and nothing is ever returned, regardless of the other arguments. However, I was thinking that it might be nice to always return the list, regardless of the `text` argument. In the case of `text = TRUE`, I think we'd want to use `invisible` so that it doesn't interfere with the printed text, but I imagine that even when folks get the text, getting everything else as a list would be useful (the text could even be added as a list item).

LewisAJones commented:
Thanks for the quick review @willgearty! I think I've addressed your comments with the latest update. This has required a bit of a shift around but should be easy enough to follow. It would still be good to eventually get all the licenses accessible via PhyloPic rather than hardcoding them!

keesey commented on R/get_attribution.R:
```diff
@@ -71,11 +81,58 @@ get_attribution <- function(uuid = NULL, img = NULL, text = FALSE) {
+                    txt)
+      # Add permalink?
+      if (permalink) {
+        txt <- paste0(txt, " Permalink: ", perm, ".")
```
Won't the `.` get interpreted as part of the URL?

LewisAJones replied:
This is just for the `text` argument which is used to generate a paragraph which users can drop into their manuscript (e.g. figure captions). If you think it is worthwhile I can drop the URL into `<URL>` or similar, but I am not sure it is necessary. Note, the actual URL is returned without `.` as an element in a list.

willgearty replied:
I think the "." is fine, but it might be better to structure the text in sentence form so that it is clear that it is a period ending the sentence rather than a symbol in the URL.

willgearty commented on R/get_attribution.R:
```diff
@@ -71,11 +81,58 @@ get_attribution <- function(uuid = NULL, img = NULL, text = FALSE) {
+        txt <- paste0(txt, " Permalink: ", perm, ".")
+      }
+    }
+    # Additional lists needed?
+    if (text || permalink) {
+      att <- list(images = att)
```
Hmm...not sure how I feel about the return behavior changing here based on these arguments. I suppose it is fine as long as this format is clear in the @return documentation, but I think it might be easiest to always have the images under an "images" list element.

willgearty commented on R/get_attribution.R:
```diff
@@ -98,40 +155,34 @@ get_attribution <- function(uuid = NULL, img = NULL, text = FALSE) {
     if (text) {
-      att <- paste0("Silhouette was contributed by ",
+      txt <- paste0("Silhouette was contributed by ",
                     att$contributor, ", ",
```
It looks like "contributors" are sometimes different from "attributors" (see https://www.phylopic.org/permalinks/33fa492ff63116d933c0c8edda57800e05e703694acda5ffe089e2cb0dd1db59). I think we want to return both of these pieces of information in the lists, but maybe only include attributors in the paragraph as is done at that permalink.

willgearty commented on R/get_attribution.R:
```diff
@@ -71,11 +81,58 @@ get_attribution <- function(uuid = NULL, img = NULL, text = FALSE) {
+    perm <- paste0("https://www.phylopic.org/permalinks/", hash)
+  }
   # API call -------------------------------------------------------------
   if (length(uuid) > 1) {
```
Hmm, on re-reviewing this, this appears to be a lot of code duplication (this `if` vs. the below `else`). Any way to reduce the code duplication?

willgearty reviewed (changes requested):
Thanks @LewisAJones. Some more suggestions and ponderings...

LewisAJones commented:
OK @willgearty, I think I've addressed all your comments now!

willgearty reviewed (approved):
Thanks @LewisAJones. This looks great, thanks for putting up with requests! I think the function is much more useful and consistent.
